op-node/p2p/discovery.go

```diff
@@ -43,8 +82,295 @@ func (conf *Config) Discovery(log log.Logger) (*enode.LocalNode, *discover.UDPv5
+	connAttempts := make(chan peer.ID, 10)
+	connectWorker := func(ctx context.Context) {
+		for {
+			id, ok := <-connAttempts
+			if !ok {
+				return
+			}
+			addrs := n.Host().Peerstore().Addrs(id)
+			log.Info("attempting connection", "peer", id)
+			ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+			err := n.Host().Connect(ctx, peer.AddrInfo{ID: id, Addrs: addrs})
+			cancel()
+			if err != nil {
+				log.Debug("failed connection attempt", "peer", id, "err", err)
+			}
+		}
+	}
+
+	// stops all the workers when we are done
+	defer close(connAttempts)
+	// start workers to try connect to peers
+	for i := 0; i < 4; i++ {
+		go connectWorker(ctx)
+	}
+
+	// buffer 10 nodes
+	randomNodesCh := make(chan *enode.Node, 3)
+	defer close(randomNodesCh)
+	bufferNodes := func() {
+		for {
+			select {
+			case <-discoverTicker.C:
+				if !randomNodeIter.Next() {
+					log.Info("discv5 DHT iteration stopped, closing peer discovery now...")
+					return
+				}
+				found := randomNodeIter.Node()
+				select {
+				// block once we have found enough nodes
+				case randomNodesCh <- found:
+					continue
+				case <-ctx.Done():
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}
+	// Walk the DHT in parallel, the discv5 interface does not use channels for the iteration
+	go bufferNodes()
+
+	// Kick off by trying the nodes we have in our table (previous nodes from last run and/or bootnodes)
+	go func() {
+		<-time.After(time.Second * 3)
+		// At the start we might have trouble walking the DHT,
+		// but we do have a table with some nodes,
+		// so take the table and feed it into the discovery process
+		for _, rec := range n.dv5Udp.AllNodes() {
+			if filter(rec) {
+				select {
+				case randomNodesCh <- rec:
+					continue
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	pstore := n.Host().Peerstore()
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("stopped peer discovery")
+			return // no ctx error, expected close
+		case found := <-randomNodesCh:
+			var dat OptimismENRData
+			if err := found.Load(&dat); err != nil { // we already filtered on chain ID and version
+				continue
+			}
+			info, err := enrToAddrInfo(found)
+			if err != nil {
+				continue
+			}
+			// We add the addresses to the peerstore, and update the address TTL to 24 hours.
+			//After that we stop using the address, assuming it may not be valid anymore (until we rediscover the node)
+			pstore.AddAddrs(info.ID, info.Addrs, time.Hour*24)
+			_ = pstore.AddPubKey(info.ID, (*crypto.Secp256k1PublicKey)(found.Pubkey()))
+			// Tag the peer, we'd rather have the connection manager prune away old peers,
+			// or peers on different chains, or anyone we have not seen via discovery.
+			// There is no tag score decay yet, so just set it to 42.
+			n.ConnectionManager().TagPeer(info.ID, fmt.Sprintf("optimism-%d-%d", dat.chainID, dat.version), 42)
+			log.Debug("discovered peer", "peer", info.ID, "nodeID", found.ID(), "addr", info.Addrs[0])
```

Comment:
can `len(info.Addrs)` ever equal zero?

Reply:
No, `enrToAddrInfo` always returns an addr-info with 1 address derived from the ENR.